### DISCUSSION TITLE
esp32 mp3 stream to analog out produces very sped up audio at 20k sample rate.

### DISCUSSION THREAD

**Discussion by eroboticdude** [pschatzmann/arduino-audio-tools#1030] [Q&A] [answered]
I have code that is using google tts url stream to play text to speech samples back through the analog output of the esp32s2, but the stream is high pitched and way too fast. Is this caused by the sample size being off? or is it an internal conversion that is going wrong?

here is my code:
```
#include <google-tts.h>
#include <WiFi.h>
#include "AudioTools.h"  // https://github.com/pschatzmann/arduino-audio-tools
#include "AudioAnalog/AnalogAudioESP32V1.h"
#include "AudioCodecs/CodecMP3MAD.h"


#include "esp32-hal-log.h"

const char* ssid = "xxxxx";
const char* password = "xxxxx";

URLStream url;  // (ssid, password);
AnalogAudioStream out;
//ResampleStream resample(out);
EncodedAudioStream dec(&out, new MP3DecoderMAD());  // Decoding stream
StreamCopy copier(dec, url);                           // copy url to decoder

TTS tts;

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  auto config = out.defaultConfig(TX_MODE);
  config.sample_rate = 15000;
  config.channels = 2;
  out.begin(config);
/*
  auto rcfg = resample.defaultConfig();
  rcfg.copyFrom(config);
  rcfg.step_size = 0.5;
  resample.begin(rcfg);
*/
  dec.setNotifyAudioChange(out);
  dec.begin();

  Serial.println("");
  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, password);

  while (WiFi.status() != WL_CONNECTED) {
    delay(250);
    Serial.print(".");
  }
  Serial.println("Wifi Connected");

  //urlAudio(tts.getSpeechUrl("Hello World, This is a test"));
}
String line;

void loop() {
  if (Serial.available() > 0) {
    line = Serial.readStringUntil('\n');
    if (line.length() < 299) {
      Serial.println(line);
      urlAudio(tts.getSpeechUrl(String(line)));
    } else {
      Serial.print("Inputted sentence is too long.");
      line = "";
    }
  }
  copier.copy();
}

void urlAudio(String UrlString) {
  url.begin(UrlString.c_str(), "audio/mp3");
  vTaskDelay(100);
}
```
I tried a resampling stream and it broke the urlstream. Not sure how. 
The audio is quite clear aside from the speed.

**Comment by pschatzmann** [comment, marked as answer]
Just added a [test example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-tts/streams-google-audiokit/streams-google-audiokit.ino) that is just using audio-tools functionality: It works perfectly for me.

You might need to upgrade if you want to use it because I needed to add the support for url encoding...

ps. If you check the log the mp3, it has a sampling rate of 24000 with 1 channel!

**Reply by eroboticdude** [reply]
yeah, i noticed that, so i set the analog sampling rate to 24000 but it was still sped up. Could it be something with the decoding stream sampling rate?

**Comment by pschatzmann** [comment]
and what did you do with the channels ?
Did you try my test sketch ?

**Reply by eroboticdude** [reply]
Sorry for the delay,
I did look at your text sketch, but it was using the I2C audio, and not the analog output. I will be moving to I2C amplifier soon, so I will let you know if it works. For now, I am having trouble with the URLStream, as it doesn't have a built-in timeout for a request. It seems to go into a lock up state when the request goes bad and I can't figure out why. Maybe the google server has a request limit per address?

**Reply by pschatzmann** [reply]
The beauty of my framework is, that you can just replace any output stream with another and it will work! 
So It does not matter that the example uses an I2SStream: just replace it with an AnalogStream or PWMOutput.

For sure Google will prevent you from abusing their infrastructure...
